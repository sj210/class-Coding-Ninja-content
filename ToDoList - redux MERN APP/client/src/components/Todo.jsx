import { useState } from "react";

import { toggleTodo, updateTodo } from "../redux/actions";
import { deleteTodo } from "../redux/actions";

import { useDispatch } from "react-redux";

const Todo = ({ todo }) => {

    const [editing, setEditing] = useState(false);
    const [text, setText] = useState(todo?.data);

    const dispatch = useDispatch();

    const onFormSubmit = (e) => {
        e.preventDefault();

        setEditing(prevState => !prevState);

        dispatch(updateTodo(todo._id, text))
    }

    return (
        <li
            className="task"
            onClick={() => dispatch(toggleTodo(todo._id))}
            style={{
                textDecoration: todo?.done ? 'line-through' : '',
                color: todo?.done ? '#bdc3c7' : '#34495e'
            }}
            data-testid="todo-test"
        >
            <span style={{ display: editing ? 'none' : '' }}>{todo?.data}</span>

            <form
                style={{ display: editing ? 'inline' : 'none' }}
                onSubmit={onFormSubmit}
            >
                <input
                    type="text"
                    value={text}
                    className="edit-todo"
                    onChange={(e) => setText(e.target.value)}
                />
            </form>

            <span className="icon" onClick={() => dispatch(deleteTodo(todo._id))}>
                <i className="fas fa-trash" />
            </span>
            <span className="icon" onClick={() => setEditing(prevState => !prevState)}>
                <i className="fas fa-pen" />
            </span>
        </li>
    )
}

export default Todo;