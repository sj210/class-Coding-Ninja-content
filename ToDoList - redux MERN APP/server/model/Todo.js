import mongoose from "mongoose";


const TodoSchema = new mongoose.Schema({
    data: {
        type: String,
        required: true
    },
    done: {
        type: Boolean,
        default: false
    },
    createdAt: {
        type: Date,
        default: Date.now
    }
})

const todo = mongoose.model('todo', TodoSchema);

export default todo;