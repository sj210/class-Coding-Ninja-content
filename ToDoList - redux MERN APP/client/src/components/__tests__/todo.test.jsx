import { render, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';

import configureStore from "redux-mock-store"
import { Provider } from 'react-redux';

import Todo from '../Todo';

afterEach(() => {
    cleanup();
})

// test('first test', () => {
//     expect(true).toBe(false);
// })

test('should render todo component', () => {
    const initialState = { state: [] };
    const mockStore = configureStore();

    const store = mockStore(initialState);

    render(
        <Provider store={store}>*
            <Todo />
        </Provider>
    );
    
    const todoElement = screen.getByTestId('todo-test');
    expect(todoElement).toBeInTheDocument();
})

test('should render our todos', () => {
    const initialState = { state: [] };
    const mockStore = configureStore();

    const store = mockStore(initialState);

    const todo = { data: 'This is a task', done: true, createdAt: ''}
    render(
        <Provider store={store}>
            <Todo todo={todo} />
        </Provider>
    );
    
    const todoElement = screen.getByTestId('todo-test');

    expect(todoElement).toHaveTextContent('This is a task');

    expect(todoElement).toContainHTML('</form>');

    expect(todoElement).not.toContainHTML('<ul>');
});


test('matches snapshot', () => {
    const todo = { data: 'This is a task', done: true, createdAt: ''}

    const initialState = { state: [] };
    const mockStore = configureStore();

    const store = mockStore(initialState);

    const tree = renderer.create(
        <Provider store={store}>
            <Todo todo={todo} />
        </Provider>
    ).toJSON();

    console.log(tree)

    expect(tree).toMatchSnapshot();
})

