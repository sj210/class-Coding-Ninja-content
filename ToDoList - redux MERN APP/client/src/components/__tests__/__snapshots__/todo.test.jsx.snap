// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<li
  className="task"
  data-testid="todo-test"
  onClick={[Function]}
  style={
    Object {
      "color": "#bdc3c7",
      "textDecoration": "line-through",
    }
  }
>
  <span
    style={
      Object {
        "display": "",
      }
    }
  >
    This is a task
  </span>
  <form
    onSubmit={[Function]}
    style={
      Object {
        "display": "none",
      }
    }
  >
    <input
      className="edit-todo"
      onChange={[Function]}
      type="text"
      value="This is a task"
    />
  </form>
  <span
    className="icon"
    onClick={[Function]}
  >
    <i
      className="fas fa-trash"
    />
  </span>
  <span
    className="icon"
    onClick={[Function]}
  >
    <i
      className="fas fa-pen"
    />
  </span>
</li>
`;
